File: hdf-converters/converters/oscal-to-hdf/go/converter_ssp.go
package oscal

import (
	"strings"

	shared "github.com/mitre/hdf-libs/hdf-converters/shared/go"
	hdf "github.com/mitre/hdf-libs/hdf-schema/dist/go"
)

// ConvertSSPToHDF converts an OSCAL System Security Plan document to an
// HDFSystem. System characteristics map to top-level system fields, components
// map to HDF Components, and control implementations are used to populate
// baseline references on components.
func ConvertSSPToHDF(input []byte, converterVersion string) (*hdf.HDFSystem, error) {
	doc, err := ParseOscalDocument(input, "system-security-plan", "oscal-ssp")
	if err != nil {
		return nil, err
	}

	ssp := doc.SystemSecurityPlan
	integrity := shared.InputIntegrity(input)

	system := &hdf.HDFSystem{
		Name:      sspSystemName(ssp),
		Integrity: integrity,
		Generator: &hdf.Generator{
			Name:    "hdf-converters",
			Version: converterVersion,
		},
	}

	// Map system characteristics
	if sc := ssp.SystemCharacteristics; sc != nil {
		if sc.Description != "" {
			desc := sc.Description
			if sc.AuthorizationBound != nil && sc.AuthorizationBound.Description != "" {
				desc += "\n\nAuthorization Boundary: " + sc.AuthorizationBound.Description
			}
			system.Description = &desc
		}

		// Map categorization level from security impact level
		if sil := sc.SecurityImpactLevel; sil != nil {
			level := sspCategorizationLevel(sil)
			system.CategorizationLevel = level
		} else if sc.SecuritySensLevel != nil && *sc.SecuritySensLevel != "" {
			level := mapSensitivityToCategorizationLevel(*sc.SecuritySensLevel)
			system.CategorizationLevel = level
		}

		// Map authorization status from system status
		if sc.Status != nil {
			status := sspAuthorizationStatus(sc.Status.State)
			system.AuthorizationStatus = status
		}

		// Map boundary description
		if sc.AuthorizationBound != nil && sc.AuthorizationBound.Description != "" {
			system.BoundaryDescription = &sc.AuthorizationBound.Description
		}

		// Map system identifier
		if len(sc.SystemIDs) > 0 {
			system.Identifier = &sc.SystemIDs[0].ID
			if sc.SystemIDs[0].IdentifierType != "" {
				system.IdentifierScheme = &sc.SystemIDs[0].IdentifierType
			}
		}
	}

	// Map version from metadata
	meta := ExtractMetadata(ssp.Metadata)
	if meta.Version != "" {
		system.Version = &meta.Version
	}

	// Build component-UUID → control-ID mapping from control-implementation
	componentControls := buildComponentControlMap(ssp.ControlImplementation)

	// Map system-implementation components to HDF Components
	if si := ssp.SystemImplementation; si != nil {
		for _, sc := range si.Components {
			comp := sspComponentToHDFComponent(&sc, componentControls)
			system.Components = append(system.Components, comp)
		}
	}

	return system, nil
}

// sspSystemName extracts the system name from SSP metadata.
func sspSystemName(ssp *SystemSecurityPlan) string {
	if sc := ssp.SystemCharacteristics; sc != nil && sc.SystemName != "" {
		return sc.SystemName
	}
	if ssp.Metadata.Title != "" {
		return ssp.Metadata.Title
	}
	return "oscal-ssp"
}

// sspCategorizationLevel determines the FIPS 199 categorization level from the
// security impact level. Uses the high water mark across CIA objectives.
func sspCategorizationLevel(sil *SecurityImpactLevel) *hdf.CategorizationLevel {
	levels := []string{sil.Confidentiality, sil.Integrity, sil.Availability}

	highest := ""
	for _, l := range levels {
		normalized := normalizeFIPSLevel(l)
		if fipsLevelRank(normalized) > fipsLevelRank(highest) {
			highest = normalized
		}
	}

	if highest == "" {
		return nil
	}

	var level hdf.CategorizationLevel
	switch highest {
	case "high":
		level = hdf.CategorizationLevelHigh
	case "moderate":
		level = hdf.Moderate
	case "low":
		level = hdf.CategorizationLevelLow
	default:
		return nil
	}
	return &level
}

// normalizeFIPSLevel normalizes FIPS 199 impact level strings.
// Handles both "fips-199-moderate" and "moderate" formats.
func normalizeFIPSLevel(level string) string {
	lower := strings.ToLower(level)
	lower = strings.TrimPrefix(lower, "fips-199-")
	switch lower {
	case "high":
		return "high"
	case "moderate", "medium":
		return "moderate"
	case "low":
		return "low"
	default:
		return ""
	}
}

// fipsLevelRank returns a numeric rank for FIPS level comparison.
func fipsLevelRank(level string) int {
	switch level {
	case "high":
		return 3
	case "moderate":
		return 2
	case "low":
		return 1
	default:
		return 0
	}
}

// mapSensitivityToCategorizationLevel maps a security-sensitivity-level string
// to an HDF CategorizationLevel.
func mapSensitivityToCategorizationLevel(level string) *hdf.CategorizationLevel {
	normalized := normalizeFIPSLevel(level)
	if normalized == "" {
		return nil
	}
	var cat hdf.CategorizationLevel
	switch normalized {
	case "high":
		cat = hdf.CategorizationLevelHigh
	case "moderate":
		cat = hdf.Moderate
	case "low":
		cat = hdf.CategorizationLevelLow
	default:
		return nil
	}
	return &cat
}

// sspAuthorizationStatus maps an OSCAL system status state to an HDF
// AuthorizationStatus.
func sspAuthorizationStatus(state string) *hdf.AuthorizationStatus {
	var status hdf.AuthorizationStatus
	switch strings.ToLower(state) {
	case "operational":
		status = hdf.Authorized
	case "under-development":
		status = hdf.PendingAuthorization
	case "disposition":
		status = hdf.Revoked
	case "other":
		status = hdf.NotYetRequested
	default:
		return nil
	}
	return &status
}

// buildComponentControlMap builds a map of component UUID → set of control IDs
// from the SSP control-implementation section.
func buildComponentControlMap(ci *SSPControlImpl) map[string]map[string]bool {
	result := make(map[string]map[string]bool)
	if ci == nil {
		return result
	}

	for _, ir := range ci.ImplementedRequirements {
		controlID := ir.ControlID

		// Direct by-components on the implemented-requirement
		for _, bc := range ir.ByComponents {
			addComponentControl(result, bc.ComponentUUID, controlID)
		}

		// by-components within statements
		for _, stmt := range ir.Statements {
			for _, bc := range stmt.ByComponents {
				addComponentControl(result, bc.ComponentUUID, controlID)
			}
		}
	}

	return result
}

// addComponentControl adds a control ID to a component's control set.
func addComponentControl(m map[string]map[string]bool, compUUID, controlID string) {
	if _, ok := m[compUUID]; !ok {
		m[compUUID] = make(map[string]bool)
	}
	m[compUUID][controlID] = true
}

// sspComponentToHDFComponent converts an OSCAL SystemComponent to an HDF
// Component, using the component-control map to populate baseline references.
func sspComponentToHDFComponent(sc *SystemComponent, componentControls map[string]map[string]bool) hdf.Component {
	comp := hdf.Component{
		Name: sc.Title,
		Type: mapOSCALComponentType(sc.Type),
	}

	if sc.Description != "" {
		comp.Description = &sc.Description
	}

	// Add control IDs as baseline refs (NIST notation)
	if controls, ok := componentControls[sc.UUID]; ok {
		var refs []string
		for controlID := range controls {
			refs = append(refs, ControlIDToNistTag(controlID))
		}
		if len(refs) > 0 {
			comp.BaselineRefs = refs
		}
	}

	return comp
}

// mapOSCALComponentType maps an OSCAL component type string to an HDF
// Copyright (component type discriminator).
func mapOSCALComponentType(oscalType string) hdf.Copyright {
	switch strings.ToLower(oscalType) {
	case "software", "this-system":
		return hdf.Application
	case "service":
		return hdf.Application
	case "hardware":
		return hdf.Host
	case "network":
		return hdf.Network
	case "database":
		return hdf.Database
	case "storage":
		return hdf.Artifact
	default:
		return hdf.Application
	}
}

File: hdf-converters/converters/oscal-to-hdf/typescript/converter-ssp.ts
/**
 * OSCAL System Security Plan (SSP) to HDF System converter.
 *
 * Mirrors the Go implementation in converters/oscal-to-hdf/go/converter_ssp.go.
 */

import { parseJSON } from '@mitre/hdf-utilities';
import { inputIntegrity, validateInputSize } from '../../../shared/typescript/converterutil.js';
import type { HDFSystem } from '@mitre/hdf-schema';
import {
  AuthorizationStatus,
  CategorizationLevel,
  TargetType,
} from '@mitre/hdf-schema';
// Import Component from system types to avoid type incompatibility with results Component
import type { Component as HdfComponent } from '@mitre/hdf-schema/hdf-system';
import type {
  Oscal,
  SystemSecurityPlanSSP,
  SecurityImpactLevel,
  AssessmentAssetsComponent,
  ControlImplementationClass,
} from './types.js';
import { controlIdToNistTag, extractMetadata } from './shared.js';

/**
 * Converts an OSCAL System Security Plan document to HDF System JSON.
 *
 * @param input - Raw JSON string containing an OSCAL SSP
 * @returns HDF System JSON string
 */
export async function convertOscalSspToHdf(input: string): Promise<string> {
  validateInputSize(input, 'oscal-ssp');

  if (!input || input.trim().length === 0) {
    throw new Error('empty input');
  }

  const doc = parseJSON<Oscal>(input);
  if (!doc['system-security-plan']) {
    throw new Error(
      "oscal-ssp: input is not a system-security-plan document (root key is not 'system-security-plan')",
    );
  }

  const ssp = doc['system-security-plan'];
  const integrity = await inputIntegrity(input);

  const system: HDFSystem = {
    name: sspSystemName(ssp),
    integrity,
    components: [],
    generator: {
      name: 'hdf-converters',
      version: '1.0.0',
    },
  };

  // Map system characteristics
  const sc = ssp['system-characteristics'];
  if (sc) {
    if (sc.description) {
      let desc = sc.description;
      if (sc['authorization-boundary']?.description) {
        desc += '\n\nAuthorization Boundary: ' + sc['authorization-boundary'].description;
      }
      system.description = desc;
    }

    // Map categorization level from security impact level
    const sil = sc['security-impact-level'];
    if (sil) {
      system.categorizationLevel = sspCategorizationLevel(sil) ?? undefined;
    } else if (sc['security-sensitivity-level']) {
      system.categorizationLevel = mapSensitivityToCategorizationLevel(sc['security-sensitivity-level']) ?? undefined;
    }

    // Map authorization status from system status
    if (sc.status) {
      system.authorizationStatus = sspAuthorizationStatus(sc.status.state) ?? undefined;
    }

    // Map boundary description
    if (sc['authorization-boundary']?.description) {
      system.boundaryDescription = sc['authorization-boundary'].description;
    }

    // Map system identifier
    const systemIds = sc['system-ids'];
    if (systemIds && systemIds.length > 0) {
      system.identifier = systemIds[0]!.id;
      if (systemIds[0]!['identifier-type']) {
        system.identifierScheme = systemIds[0]!['identifier-type'];
      }
    }
  }

  // Map version from metadata
  const meta = extractMetadata(ssp.metadata);
  if (meta.version) {
    system.version = meta.version;
  }

  // Build component-UUID -> control-ID mapping
  const componentControls = buildComponentControlMap(ssp['control-implementation']);

  // Map system-implementation components to HDF Components
  const si = ssp['system-implementation'];
  if (si?.components) {
    for (const comp of si.components) {
      system.components.push(sspComponentToHDFComponent(comp, componentControls));
    }
  }

  return JSON.stringify(system, null, 2);
}

function sspSystemName(ssp: SystemSecurityPlanSSP): string {
  const sc = ssp['system-characteristics'];
  if (sc?.['system-name']) {
    return sc['system-name'];
  }
  if (ssp.metadata.title) {
    return ssp.metadata.title;
  }
  return 'oscal-ssp';
}

function sspCategorizationLevel(
  sil: SecurityImpactLevel,
): CategorizationLevel | null {
  const levels = [
    sil['security-objective-confidentiality'] ?? '',
    sil['security-objective-integrity'] ?? '',
    sil['security-objective-availability'] ?? '',
  ];

  let highest = '';
  for (const l of levels) {
    const normalized = normalizeFIPSLevel(l);
    if (fipsLevelRank(normalized) > fipsLevelRank(highest)) {
      highest = normalized;
    }
  }

  if (!highest) return null;

  switch (highest) {
    case 'high':
      return CategorizationLevel.High;
    case 'moderate':
      return CategorizationLevel.Moderate;
    case 'low':
      return CategorizationLevel.Low;
    default:
      return null;
  }
}

function normalizeFIPSLevel(level: string): string {
  let lower = level.toLowerCase();
  lower = lower.replace(/^fips-199-/, '');
  switch (lower) {
    case 'high':
      return 'high';
    case 'moderate':
    case 'medium':
      return 'moderate';
    case 'low':
      return 'low';
    default:
      return '';
  }
}

function fipsLevelRank(level: string): number {
  switch (level) {
    case 'high':
      return 3;
    case 'moderate':
      return 2;
    case 'low':
      return 1;
    default:
      return 0;
  }
}

function mapSensitivityToCategorizationLevel(
  level: string,
): CategorizationLevel | null {
  const normalized = normalizeFIPSLevel(level);
  if (!normalized) return null;

  switch (normalized) {
    case 'high':
      return CategorizationLevel.High;
    case 'moderate':
      return CategorizationLevel.Moderate;
    case 'low':
      return CategorizationLevel.Low;
    default:
      return null;
  }
}

function sspAuthorizationStatus(state: string): AuthorizationStatus | null {
  switch (state.toLowerCase()) {
    case 'operational':
      return AuthorizationStatus.Authorized;
    case 'under-development':
      return AuthorizationStatus.PendingAuthorization;
    case 'disposition':
      return AuthorizationStatus.Revoked;
    case 'other':
      return AuthorizationStatus.NotYetRequested;
    default:
      return null;
  }
}

function buildComponentControlMap(
  ci: ControlImplementationClass | undefined,
): Map<string, Set<string>> {
  const result = new Map<string, Set<string>>();
  if (!ci) return result;

  for (const ir of ci['implemented-requirements'] ?? []) {
    const controlId = ir['control-id'];

    // Direct by-components on the implemented-requirement
    for (const bc of ir['by-components'] ?? []) {
      addComponentControl(result, bc['component-uuid'], controlId);
    }

    // by-components within statements
    for (const stmt of ir.statements ?? []) {
      for (const bc of stmt['by-components'] ?? []) {
        addComponentControl(result, bc['component-uuid'], controlId);
      }
    }
  }

  return result;
}

function addComponentControl(
  m: Map<string, Set<string>>,
  compUUID: string,
  controlId: string,
): void {
  let controls = m.get(compUUID);
  if (!controls) {
    controls = new Set<string>();
    m.set(compUUID, controls);
  }
  controls.add(controlId);
}

function sspComponentToHDFComponent(
  sc: AssessmentAssetsComponent,
  componentControls: Map<string, Set<string>>,
): HdfComponent {
  const comp: HdfComponent = {
    name: sc.title,
    type: mapOSCALComponentType(sc.type),
  };

  if (sc.description) {
    comp.description = sc.description;
  }

  // Add control IDs as baseline refs (NIST notation)
  const controls = componentControls.get(sc.uuid);
  if (controls && controls.size > 0) {
    const refs: string[] = [];
    for (const controlId of controls) {
      refs.push(controlIdToNistTag(controlId));
    }
    if (refs.length > 0) {
      comp.baselineRefs = refs;
    }
  }

  return comp;
}

function mapOSCALComponentType(oscalType: string): TargetType {
  switch (oscalType.toLowerCase()) {
    case 'software':
    case 'this-system':
      return TargetType.Application;
    case 'service':
      return TargetType.Application;
    case 'hardware':
      return TargetType.Host;
    case 'network':
      return TargetType.Network;
    case 'database':
      return TargetType.Database;
    case 'storage':
      return TargetType.Artifact;
    default:
      return TargetType.Application;
  }
}
